rcore-fs-sefs/src/lib.rs

```diff
@@ -207,6 +211,149 @@ impl INodeImpl {
+            (start, end - start)
+        } else {
+            (range.start, range.len())
+        };
+        self.file
+            .write_zeros_at(offset, len)
+            .expect("failed to write zeros");
+
+        // May update file size
+        if !keep_size && range.end > file_size {
+            self.file.set_len(range.end)?;
+            inode.size = range.end as u64;
+        }
+        Ok(())
+    }
+
+    /// Insert zeros for the specific range of file, the contents of the file
+    /// starting at the start of range will be shifted upward, the file will be
+    /// range.len() bytes bigger.
+    fn insert_range(&self, range: &Range<usize>) -> vfs::Result<()> {
+        let mut inode = self.disk_inode.write();
+        let file_size = inode.size as usize;
+        if range.start >= file_size {
+            return Err(FsError::InvalidParam);
+        }
+        if range.start % BLKSIZE != 0 || range.len() % BLKSIZE != 0 {
+            return Err(FsError::InvalidParam);
+        }
+
+        // Move the contents of file forward
+        let old_range = Range::<usize> {
+            start: range.start,
+            end: file_size,
+        };
+        self.copy_range(&old_range, range.len(), &CopyRangeDirection::Forward)?;
+
+        // Insert zeros
+        self.file
+            .write_zeros_at(range.start, range.len())
+            .expect("failed to write zeros");
+
+        // Update file size
+        let new_file_size = file_size + range.len();
+        self.file.set_len(new_file_size)?;
+        inode.size = new_file_size as u64;
+        Ok(())
+    }
+
+    /// Remove the range of file, the contents of the file starting at the end
+    /// of range will be appended at the start of the range, the file will be
+    /// range.len() bytes smaller.
+    fn collapse_range(&self, range: &Range<usize>) -> vfs::Result<()> {
+        let mut inode = self.disk_inode.write();
+        let file_size = inode.size as usize;
+        if range.end >= file_size {
+            return Err(FsError::InvalidParam);
+        }
+        if range.start % BLKSIZE != 0 || range.len() % BLKSIZE != 0 {
+            return Err(FsError::InvalidParam);
+        }
+
+        // Move the contents of file backward
+        let old_range = Range::<usize> {
+            start: range.end,
+            end: file_size,
+        };
+        self.copy_range(&old_range, range.len(), &CopyRangeDirection::Backward)?;
+
+        // Update file size
+        let new_file_size = file_size - range.len();
+        self.file.set_len(new_file_size)?;
+        inode.size = new_file_size as u64;
+        Ok(())
+    }
+
+    /// Copy the range of file `offset` bytes backward or forward, this functions
+    /// can only be called by `insert_range` or `collapse_range`.
+    fn copy_range(
+        &self,
+        range: &Range<usize>,
+        offset: usize,
+        direction: &CopyRangeDirection,
```

Comment:
The third argument is unnecessary. Copying a range of data without shooting ourselves in the foot---mistakenly overriding data before copying---is simple: if the `offset < range.start`, coping blocks from low addresses to high ones; otherwise, copying from high ones to low.

In addition, this API signature looks ambiguous. I have to look at the user code (`insert_range` and `collapse_range`) before getting a clue about the exact meaning of this API.

See how the following version 1) clears up the intention of the API, 2) make the logic more clear, and 3) shorter. (I didn't compile or test it. So it could be buggy. But it should give you a sense of the direction for improvement).

```rust
/// Copy a range of data to a specified offset. The source range and the destination range may overlap.
fn copy_range_to(&self, src_range: &Range<usize>, dst_offset: usize) -> vfs::Result<()> {
    // Choose the copying direction to avoid overriding data beforing copying
    let (mut src_offset, mut dst_offset, mut offset_step) = if src_range.start < dst_offset {
        let src_offset = src_range.end - BLKSIZE;
        let dst_offset = dst_offset + src_range.len() - BLKSIZE;
        let offset_step = -BLKSIZE;
        (src_offset, dst_offset, offset_step)
    } else if src_range.start > dst_offset {
        let read_offset = src_range.start;
        let write_offset = dst_offset;
        let offset_step = BLKSIZE;
        (src_offset, dst_offset, offset_step)
    } else {
        return Ok(());
    };

    // Do the copying
    let mut remain_size = src_range.len();
    let mut buf = unsafe {
        MaybeUninit::<[u8; BLKSIZE]>::new_uninit().assume_init();
    };
    while remain_size > 0 {
        let len = remain_size.min(BLKSIZE);
        self.file
            .read_exact_at(&mut buf[..len], src_offset)
            .unwrap();
        self.file.write_all_at(&buf[..len], dst_offset).unwrap();
        remain_size -= len;
        src_offset += offset_step;
        dst_offset += offset_step;
    }
    Ok(())
}
```